dataset.py

```diff
@@ -139,6 +182,226 @@ def _partition_data_random(sample_data: pd.DataFrame,
+  and finally test_coord.
+  '''
+  train_split = _nearest_neighbors(
+    center_coordinate=train_coord,
+    sample_data=sample_data,
+    n_neighbors=sample_data.shape[0]*strategy.train_fraction
+  )
+  filtered_sample_data = sample_data[
+          ~sample_data.index.isin(
+            train_split.index.values
+          )]
+  
+  validation_split = _nearest_neighbors(
+    center_coordinate=validation_coord,
+    sample_data=filtered_sample_data,
+    n_neighbors=sample_data.shape[0]*strategy.validation_fraction 
+  )
+  filtered_sample_data = filtered_sample_data[
+    ~filtered_sample_data.index.isin(
+      validation_split.index.values
+    )]
+  
+  test_split = _nearest_neighbors(
+    center_coordinate=test_coord,
+    sample_data=filtered_sample_data,
+    n_neighbors=filtered_sample_data.shape[0] 
+  )
+
+  return PartitionedDataset(
+    train=train_split,
+    test=test_split,
+    validation=validation_split
+  )
+
+def _polygon(
+  coordinates: list[list]) -> Polygon:
+  '''
+  Given an unordered list of longitude and latitude coordinates
+  represented by tuples, returns a polygon with the shape of the
+  coordinates.
+  '''
+  if len(coordinates) < 3:
+    return None
+  # Compute centroid
+  centroid = (sum([c[0] for c in coordinates])/len(coordinates),
+              sum([c[1] for c in coordinates])/len(coordinates))
+  # Sort by polar angle. This arranges the coordinates so that they
+  # delimit an outward shape.
+  sorted_coordinates = list(coordinates)
+  sorted_coordinates.sort(
+    key=lambda c: math.atan2(
+      c[1] - centroid[1],
+      c[0] - centroid[0])
+  )
+  return Polygon(sorted_coordinates)
+
+def _valid_polygons(
+  train_polygon: Polygon,
+  validation_polygon: Polygon,
+  test_polygon: Polygon) -> bool:
+  '''
+  Returns true if the train, validation, and test polygons
+  have a larger area than or equal to 0.1 and they don't overlap
+  over each other. Returns false otherwise. 
+  '''
+  small_polygon = False
+  for polygon in [train_polygon, validation_polygon, test_polygon]:
+    if polygon.area < 0.1:
+      return False
+
+  if (train_polygon.intersects(validation_polygon) or
+     train_polygon.intersects(test_polygon) or
+     validation_polygon.intersects(test_polygon)):
+    return False
+
+  return True
+
+def _shuffled_unique_coordinates(
+  sample_data: pd.DataFrame,
+  strategy: FurthestPointsPartitionStrategy) -> list[list[float]]:
+  '''
+  Returns shuffled unique coordinates from sample_data that match
+  the expected train, validation and test fractions.
+  '''
+  unique_coordinates_df = sample_data.groupby(
+    by=['long', 'lat']
+  ).first().reset_index()
+  shuffled_coordinates_df = unique_coordinates_df.sample(frac=1.0,
+    random_state=strategy.random_seed)
+  coordinates = list(
+    shuffled_coordinates_df[
+      [_LONGITUDE_COLUMN_NAME, _LATITUDE_COLUMN_NAME]
+    ].values)
+  assert(len(coordinates) >= strategy.top_n)
```

Comment:
These asserts should probably be some kind of an issue-specific exception. I generally don't see this in libraries. Here's one stackoverflow post that discusses this: https://stackoverflow.com/questions/3674962/assert-vs-exception-when-developing-the-library

Exceptions can provide more information about what's going wrong and it's just more of a standard way to validate API parameters. I guess asserts here could make sense if the external-facing part of this API has no chance of triggering them unless we did something wrong. Is that the case here?

Reply:
Our internal style guidelines are even more prescriptive than this, but we have a good amount of flexibility here since this is external code.